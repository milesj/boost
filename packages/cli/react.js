module.exports = require('./cjs/react.cjs');
