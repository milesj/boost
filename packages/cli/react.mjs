export * from './mjs/react.mjs';
