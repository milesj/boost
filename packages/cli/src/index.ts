/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './CLIError';
export * from './Command';
export * from './constants';
export { Config } from './decorators/Config';
export * from './helpers';
export * as Arg from './metadata/args';
export * from './middleware';
export * from './Program';
export * from './types';
