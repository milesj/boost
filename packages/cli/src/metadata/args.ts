export * from '../decorators';
export * from '../initializers';
