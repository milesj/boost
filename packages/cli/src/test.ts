/* eslint-disable no-param-reassign */

import React from 'react';
import { render } from 'ink';
import { env } from '@boost/internal';
import { mockLogger } from '@boost/log/test';
import { stripAnsi } from '@boost/terminal';
import { Command, INTERNAL_OPTIONS, INTERNAL_PARAMS, Program } from '.';

import type {
	ExitCode,
	GlobalOptions,
	PrimitiveType,
	ProgramOptions,
	ProgramStreams,
	TaskContext,
} from '.';

export class MockReadStream {
	isTTY = false;
}

export class MockWriteStream {
	append: boolean = false;

	columns: number = 80;

	output: string;

	constructor(append: boolean = false) {
		this.append = append;
		this.output = '';
	}

	write(string: string) {
		if (this.append) {
			this.output += string;
		} else {
			this.output = string;
		}
	}

	get(): string {
		return this.output;
	}

	on() {}

	off() {}
}

/**
 * Returns mocked `stderr`, `stdout`, and `stdin` streams that can be passed to a `Program`.
 * This does not mock all stream functionality, only those required by Boost and Ink.
 *
 * ```ts
 * import { mockStreams } from '@boost/cli/test';
 *
 * const streams = mockStreams();
 * ```
 */
export function mockStreams(append?: boolean): ProgramStreams {
	return {
		stderr: new MockWriteStream(append),
		stdin: new MockReadStream(),
		stdout: new MockWriteStream(append),
	} as unknown as ProgramStreams;
}

/**
 * Returns a `Program` instance with required options pre-filled and streams mocked
 * (unless manually provided).
 *
 * ```ts
 * import { mockProgram } from '@boost/cli/test';
 *
 * const program = mockProgram({ name: 'Example' });
 * ```
 */
export function mockProgram(options?: Partial<ProgramOptions>, streams?: ProgramStreams): Program {
	return new Program(
		{
			bin: 'test',
			name: 'Test',
			version: '0.0.0',
			...options,
		},
		streams ?? mockStreams(),
	);
}

/**
 * Can be used to render a React component with Ink and return the rendered result
 * as a terminal compatible string. If `stripped` is true, it will strip ANSI
 * escape escape sequences.
 *
 * ```tsx
 * import { renderComponent } from '@boost/cli/test';
 * import TestComponent from '../src/components/TestComponent';
 *
 * it('renders a component', async () => {
 * 	expect(await renderComponent(<TestComponent />)).toMatchSnapshot();
 * });
 * ```
 *
 * > As an alternative, we also suggest using the official
 * > [ink-testing-library](https://github.com/vadimdemedes/ink-testing-library).
 */
export async function renderComponent(
	element: React.ReactElement,
	stripped: boolean = false,
): Promise<string> {
	let output = '';

	try {
		const { render: renderAsTest } = await import('ink-testing-library');
		const { lastFrame } = await renderAsTest(element);

		output = lastFrame() ?? '';
	} catch {
		const stdout = new MockWriteStream();

		await render(element, {
			debug: true,
			experimental: true,
			stdout: stdout as unknown as NodeJS.WriteStream,
		});

		output = stdout.get();
	}

	return stripped ? stripAnsi(output) : output;
}

/**
 * Runs a `Command` outside the context of a `Program`, but mimics similar functionality,
 * including React component rendering. Params are required as they're passed to the run method,
 * while options are optional and assume class properties have been defined. Also, the `exit`
 * and `log` methods have been mocked with Vitest spies so that they can be asserted.
 *
 * ```ts
 * import { runCommand } from '@boost/cli/test';
 * import TestCommand from '../src/commands/TestCommand';
 *
 * it('runs a command', async () => {
 * 	const command = new TestCommand();
 *
 * 	expect(await runCommand(command, ['foo', 'bar', 'baz'])).toMatchSnapshot();
 * 	expect(command.log).toHaveBeenCalled();
 * });
 * ```
 *
 * > Since there is no `Program` context, any functionality that requires a program will fail. If
 * > needed, use `runProgram()` instead.
 */
export async function runCommand<O extends GlobalOptions, P extends PrimitiveType[]>(
	command: Command<O, P>,
	params: P,
	options?: Partial<O>,
): Promise<string> {
	const { vi } = await import('vitest');

	if (options) {
		Object.assign(command, options);

		// @ts-expect-error Allow overwrite
		command[INTERNAL_OPTIONS] = {
			help: false,
			locale: 'en',
			version: false,
			...options,
		};
	}

	command.exit = vi.fn();
	// eslint-disable-next-line require-atomic-updates
	command.log = await mockLogger();
	// eslint-disable-next-line require-atomic-updates
	command[INTERNAL_PARAMS] = params;

	const result = await command.run(...params);

	if (!result || typeof result === 'string') {
		return result ?? '';
	}

	return renderComponent(result);
}

/**
 * Runs a task function outside the context of a `Command`, in complete isolation.
 * A mock command context is provided with standard defaults, and can be customized
 * through the 3rd argument.
 *
 * ```ts
 * import { runTask } from '@boost/cli/test';
 * import testTask from '../src/tasks/testTask';
 *
 * it('runs a task', async () => {
 * 	const context = {
 * 		log: vi.fn(),
 * 	};
 *
 * 	expect(await runTask(testTask, ['foo', 'bar', 'baz'], context)).toMatchSnapshot();
 * 	expect(context.log).toHaveBeenCalled();
 * });
 * ```
 */
export async function runTask<A extends unknown[], R, T extends TaskContext>(
	task: (this: T, ...argz: A) => R,
	args: A,
	context?: Partial<T>,
): Promise<R> {
	const { vi } = await import('vitest');
	const notTestable = (name: string) => () => {
		throw new Error(
			`\`${name}\` is not testable using the \`runTask\` utility. Test using a full program.`,
		);
	};

	const baseContext: TaskContext = {
		exit: vi.fn(),
		help: false,
		locale: 'en',
		log: await mockLogger(),
		rest: [] as string[],
		runProgram: notTestable('runProgram'),
		runTask: notTestable('runTask'),
		unknown: {},
		version: false,
	};

	return task.apply(
		{
			...baseContext,
			...context,
		} as T,
		args,
	);
}

/**
 * Runs a `Program` as if it were ran on the command line, including middleware,
 * commands, success and failure states, and more. Utilizes mocked streams to capture
 * and return standard output and ANSI stripped output. Failed runs will not throw
 * and instead will render a failure output.
 *
 * ```ts
 * import { runProgram } from '@boost/cli/test';
 * import Program from '../src/program';
 *
 * it('runs a program', async () => {
 * 	const program = new Program();
 *
 * 	const { code, output } = await runProgram(program, ['cmd', '--foo', '123', 'bar']);
 *
 * 	expect(output).toMatchSnapshot();
 * 	expect(code).toBe(0);
 * });
 * ```
 */
export async function runProgram(
	program: Program,
	argv: string[],
	options: { append?: boolean } = {},
): Promise<{ code: ExitCode; output: string; outputStripped: string }> {
	if (!(program.streams.stderr instanceof MockWriteStream)) {
		program.streams.stderr = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdout instanceof MockWriteStream)) {
		program.streams.stdout = new MockWriteStream(options.append) as unknown as NodeJS.WriteStream;
	}

	if (!(program.streams.stdin instanceof MockReadStream)) {
		program.streams.stdin = new MockReadStream() as unknown as NodeJS.ReadStream;
	}

	// Ink async rendering never resolves while testing,
	// as it relies on system signals to "exit".
	// So we set this to flag the renderer to avoid awaiting.
	env('CLI_TEST_ONLY', 'true');

	const code = await program.run(argv);

	env('CLI_TEST_ONLY', null);

	const output =
		(program.streams.stdout as unknown as MockWriteStream).get() +
		(program.streams.stderr as unknown as MockWriteStream).get();

	return {
		code,
		output,
		outputStripped: stripAnsi(output),
	};
}
