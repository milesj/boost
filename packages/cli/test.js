const { Command } = require('./lib');

const c = new Command();
