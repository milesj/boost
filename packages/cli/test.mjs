export * from './mjs/test.mjs';
