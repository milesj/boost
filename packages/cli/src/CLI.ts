import { Argv } from '@boost/args';
import { Contract, Predicates } from '@boost/common';
import { CLIOptions } from './types';

export default class CLI extends Contract<CLIOptions> {
  blueprint({ string }: Predicates) {
    return {
      bin: string().required(),
      name: string().required(),
      version: string().required(),
    };
  }

  run(argv: Argv) {}
}
