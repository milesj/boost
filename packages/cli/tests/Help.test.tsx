import React from 'react';
import { describe, expect,it, vi } from 'vitest';
import { Help } from '../src/react';
import { renderComponent } from '../src/test';
import { commands, options, params } from './__fixtures__/args';

vi.mock('term-size');

describe('<Help />', () => {
	it('renders everything', async () => {
		await expect(
			renderComponent(
				<Help
					commands={commands}
					config={{ description: 'I am a command that does cool things.', usage: '$ ink foo bar' }}
					options={options}
					params={params}
				/>,
			),
		).resolves.toMatchSnapshot();
	});

	describe('config', () => {
		it('renders description', async () => {
			await expect(
				renderComponent(<Help config={{ description: 'I am a command that does cool things.' }} />),
			).resolves.toMatchSnapshot();
		});

		it('renders description with additional params', async () => {
			await expect(
				renderComponent(
					<Help
						config={{ description: 'I am a command that does cool things.', deprecated: true }}
					/>,
				),
			).resolves.toMatchSnapshot();
		});

		it('renders description with markdown', async () => {
			await expect(
				renderComponent(
					<Help
						config={{
							description:
								'This is the top level command description.\nAll descriptions support markdown like **bold**, __strong__, *italics*, _emphasis_, and ~~strikethroughs~~.',
						}}
					/>,
				),
			).resolves.toMatchSnapshot();
		});
	});

	describe('usage', () => {
		it('renders usage string', async () => {
			await expect(
				renderComponent(<Help config={{ description: '', usage: '$ ink foo bar' }} />),
			).resolves.toMatchSnapshot();
		});

		it('renders usage array', async () => {
			await expect(
				renderComponent(
					<Help config={{ description: '', usage: ['$ ink foo bar', '$ test --foo -b'] }} />,
				),
			).resolves.toMatchSnapshot();
		});
	});

	describe('params', () => {
		it('doesnt render empty params', async () => {
			await expect(renderComponent(<Help params={[]} />)).resolves.toMatchSnapshot();
		});

		it('renders params', async () => {
			await expect(renderComponent(<Help params={params} />)).resolves.toMatchSnapshot();
		});

		it('renders params (stripped)', async () => {
			await expect(renderComponent(<Help params={params} />, true)).resolves.toMatchSnapshot();
		});
	});

	describe('commands', () => {
		it('doesnt render empty commands', async () => {
			await expect(renderComponent(<Help commands={{}} />)).resolves.toMatchSnapshot();
		});

		it('renders commands', async () => {
			await expect(renderComponent(<Help commands={commands} />)).resolves.toMatchSnapshot();
		});

		it('renders commands (stripped)', async () => {
			await expect(renderComponent(<Help commands={commands} />, true)).resolves.toMatchSnapshot();
		});
	});

	describe('options', () => {
		it('doesnt render empty options', async () => {
			await expect(renderComponent(<Help options={{}} />)).resolves.toMatchSnapshot();
		});

		it('renders options', async () => {
			await expect(renderComponent(<Help options={options} />)).resolves.toMatchSnapshot();
		});

		it('renders options (stripped)', async () => {
			await expect(renderComponent(<Help options={options} />, true)).resolves.toMatchSnapshot();
		});
	});
});
