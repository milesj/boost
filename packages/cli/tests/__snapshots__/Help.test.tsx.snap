// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Help /> commands doesnt render empty commands 1`] = `""`;

exports[`<Help /> commands renders commands (stripped) 1`] = `
"
 COMMANDS 

 barbar         This is a deprecated command with no params. [deprecated]
 foo src [dst]  This is a normal command with params."
`;

exports[`<Help /> commands renders commands 1`] = `
"
[1m[47m[30m COMMANDS [39m[49m[22m

 barbar         This is a deprecated command with no params. [90m[deprecated][39m
 foo [90msrc[39m [90m[dst][39m  This is a normal command with params."
`;

exports[`<Help /> config renders description 1`] = `" I am a command that does cool things."`;

exports[`<Help /> config renders description with additional params 1`] = `" I am a command that does cool things. [90m[deprecated][39m"`;

exports[`<Help /> config renders description with markdown 1`] = `
" This is the top level command description.
 All descriptions support markdown like [1mbold[22m, [1mstrong[22m, [3mitalics[23m, [3memphasis[23m, and [9mstrikethroughs[29m."
`;

exports[`<Help /> options doesnt render empty options 1`] = `""`;

exports[`<Help /> options renders options (stripped) 1`] = `
"
 OPTIONS 

     --deprecatedFlag                        A deprecated flag. [deprecated]
     --deprecatedStringWithALongName string  Deprecated string option with custom label.
                                             [deprecated]
     --flag                                  Standard boolean option.
     --num number                            Standard number option.
 -C, --numberCountable number                Number option that increments a count for each
                                             occurence. (default: 10) [counter]
 -n, --numberFixedList number                A list of numbers to choose from. (choices: 1, 2, 3, 4,
                                              5, 6, 7, 8, 9, 10) [deprecated]
     --numWithArity number[]                 A multiple number option that requires an exact number
                                             of values. [multiple(10)]
     --str string                            Standard string option.
 -s, --stringChoices string                  Choice list with strings. (choices: \\"foo\\", \\"bar\\",
                                             \\"baz\\") (default: \\"bar\\")
 -S, --strList string[]                      String option list. [multiple]
 -t, --no-truthyBool                         Flag that is on by default, so name should be negated.
                                             (default: true)"
`;

exports[`<Help /> options renders options 1`] = `
"
[1m[47m[30m OPTIONS [39m[49m[22m

     --deprecatedFlag                        A deprecated flag. [90m[deprecated][39m
     --deprecatedStringWithALongName [90mstring[39m  Deprecated string option with custom label.
                                             [90m[deprecated][39m
     --flag                                  Standard boolean option.
     --num [90mnumber[39m                            Standard number option.
 -C, --numberCountable [90mnumber[39m                Number option that increments a count for each
                                             occurence. (default: 10) [90m[counter][39m
 -n, --numberFixedList [90mnumber[39m                A list of numbers to choose from. (choices: 1, 2, 3, 4,
                                              5, 6, 7, 8, 9, 10) [90m[deprecated][39m
     --numWithArity [90mnumber[][39m                 A multiple number option that requires an exact number
                                             of values. [90m[multiple(10)][39m
     --str [90mstring[39m                            Standard string option.
 -s, --stringChoices [90mstring[39m                  Choice list with strings. (choices: \\"foo\\", \\"bar\\",
                                             \\"baz\\") (default: \\"bar\\")
 -S, --strList [90mstring[][39m                      String option list. [90m[multiple][39m
 -t, --no-truthyBool                         Flag that is on by default, so name should be negated.
                                             (default: true)"
`;

exports[`<Help /> params doesnt render empty params 1`] = `""`;

exports[`<Help /> params renders params (stripped) 1`] = `
"
 PARAMS 

 0 string     Complex param with many tags. [deprecated, required]
 path string  This is a string param thats required. [required]
 ms number    And this a number param with a label.
 3 boolean    And finally a boolean param. [deprecated]"
`;

exports[`<Help /> params renders params 1`] = `
"
[1m[47m[30m PARAMS [39m[49m[22m

 0 [90mstring[39m     Complex param with many tags. [90m[deprecated, required][39m
 path [90mstring[39m  This is a string param thats required. [90m[required][39m
 ms [90mnumber[39m    And this a number param with a label.
 3 [90mboolean[39m    And finally a boolean param. [90m[deprecated][39m"
`;

exports[`<Help /> renders everything 1`] = `
" I am a command that does cool things.

[1m[47m[30m USAGE [39m[49m[22m

 $ ink foo bar

[1m[47m[30m PARAMS [39m[49m[22m

 0 [90mstring[39m     Complex param with many tags. [90m[deprecated, required][39m
 path [90mstring[39m  This is a string param thats required. [90m[required][39m
 ms [90mnumber[39m    And this a number param with a label.
 3 [90mboolean[39m    And finally a boolean param. [90m[deprecated][39m

[1m[47m[30m COMMANDS [39m[49m[22m

 barbar         This is a deprecated command with no params. [90m[deprecated][39m
 foo [90msrc[39m [90m[dst][39m  This is a normal command with params.

[1m[47m[30m OPTIONS [39m[49m[22m

     --deprecatedFlag                        A deprecated flag. [90m[deprecated][39m
     --deprecatedStringWithALongName [90mstring[39m  Deprecated string option with custom label.
                                             [90m[deprecated][39m
     --flag                                  Standard boolean option.
     --num [90mnumber[39m                            Standard number option.
 -C, --numberCountable [90mnumber[39m                Number option that increments a count for each
                                             occurence. (default: 10) [90m[counter][39m
 -n, --numberFixedList [90mnumber[39m                A list of numbers to choose from. (choices: 1, 2, 3, 4,
                                              5, 6, 7, 8, 9, 10) [90m[deprecated][39m
     --numWithArity [90mnumber[][39m                 A multiple number option that requires an exact number
                                             of values. [90m[multiple(10)][39m
     --str [90mstring[39m                            Standard string option.
 -s, --stringChoices [90mstring[39m                  Choice list with strings. (choices: \\"foo\\", \\"bar\\",
                                             \\"baz\\") (default: \\"bar\\")
 -S, --strList [90mstring[][39m                      String option list. [90m[multiple][39m
 -t, --no-truthyBool                         Flag that is on by default, so name should be negated.
                                             (default: true)"
`;

exports[`<Help /> usage renders usage array 1`] = `
"
[1m[47m[30m USAGE [39m[49m[22m

 $ ink foo bar
 $ test --foo -b"
`;

exports[`<Help /> usage renders usage string 1`] = `
"
[1m[47m[30m USAGE [39m[49m[22m

 $ ink foo bar"
`;
